HACKING.md

```diff
@@ -247,6 +244,28 @@ In a Debian/Ubuntu build environment you can build debs with
 
     tools/make-debs --quick
 
+## Updating `node_modules`
+
+During a normal build from a git checkout, the `node_modules` will be
+automatically unpacked from a cache kept in a separate git repository.  You can
+force the unpack to occur using the `tools/node-modules checkout` command, but
+this shouldn't be necessary.  In the event that you need to modify
+`package.json` (to install a new module, for example) then you'll need to run
+`tools/node-modules rebuild` to create a new cache from the result of running
+`npm install` on your new `package.json`.
+
+Before landing changes to the main cockpit project (or proposing a PR), this
+rebuild will need to be pushed to the cache repository.  `tools/node-modules
+push` will do this for you.
```

Comment:
This feels conceptually wrong to me. A PR can get rejected, superseded, rebased, and contributors don't even have the permission to push to the cache. We will somehow need to find a way to do this on the CI side.

Reply:
I disagree.  I keep comparing this to image refreshes, because I think it's a very good comparison.  Bots (or even humans) push images to the mirrors, and then we update a symlink (in this case gitlink) in the cockpits repo and propose a PR and do review there.  We can reject broken vm images, and we can reject broken node_modules updates in exactly the same way.  Also note that the node_modules commits are completely independent of each other:  There is no history, only tags to single parentless commits...

The only thing missing here is a prune.

I guess you imagine some kind of a procedure on a PR to have a bot come along and turn package.json into a commit in our `node-cache` repository?  This might be helpful, but it's also annoying: it means that if you're hacking on something that requires a change to package.json, you'd first need to create a PR, push your package.json change, ask the bot to do the build, pull the bot's commit into your local repo, and then continue working.... :/

Reply:
I had assumed that the classic `npm install` would still work locally. If that's not the case, then we can try this, but I still don't like that it divides PR contributors into two classes. (this is why I really like the dist caching -- it doesn't get in the way and doesn't make workflows more complicated). I think on IRC you mentioned something like that, like computing a hash over package-lock, and using that as lookup value.

Anyway, I don't want this to be a hard veto -- we can try it, and if it doesn't work for us, go back.

Reply:
image refreshes  are actually a good comparison, right -- I can change the scripts and run image-create locally (the equivalent of `npm install`), and then can develop against this right away. There is no need to first push this somewhere and let the bot commit anything. But AFAICS this would work here, right? For locally running webpack/image-prepare etc. I wouldn't *have* to push, I can first test this locally if it's even something that I want, right?

Reply:
Exactly correct.

But as is the case with image refreshes, if you want *other* people to be able to use it (and in any case, before landing a PR) you'll need to publish it.